src/net/neighbors.rs

```diff
@@ -0,0 +1,2557 @@
+                        }
+
+                        self.set_state(local_peer, NeighborWalkState::GetHandshakesBegin);
+                        Ok(Some(to_resolve))
+                    },
+                    StacksMessageType::Nack(ref data) => {
+                        info!("Neighbor {:?} NACK'ed GetNeighbors with code {:?}", &self.cur_neighbor.addr, data.error_code);
+                        self.result.add_broken(self.cur_neighbor.addr.clone());
+                        Err(net_error::ConnectionBroken)
+                    },
+                    _ => {
+                        // invalid message
+                        info!("Got out-of-sequence message from {:?}", &self.cur_neighbor.addr);
+                        self.result.add_broken(self.cur_neighbor.addr.clone());
+                        Err(net_error::InvalidMessage)
+                    }
+                }
+            },
+            Err(req_res) => {
+                match req_res {
+                    Ok(same_req) => {
+                        // try again
+                        self.getneighbors_request = Some(same_req);
+                        Ok(None)
+                    },
+                    Err(e) => {
+                        // disconnected 
+                        self.result.add_broken(self.cur_neighbor.addr.clone());
+                        Err(e)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Begin getting the neighors of cur_neighbor's neighbors.
+    /// NetworkReplyHandles should be reply handles for Handshake requests.
+    pub fn neighbor_handshakes_begin(&mut self, local_peer: &LocalPeer, mut handshake_handles: HashMap<NeighborAddress, NetworkReplyHandle>) -> () {
+        assert!(self.state == NeighborWalkState::GetHandshakesBegin);
+
+        // advance state!
+        self.unresolved_handshake_neighbors.clear();
+        for (naddr, nh) in handshake_handles.drain() {
+            self.unresolved_handshake_neighbors.insert(naddr, nh);
+        }
+
+        self.set_state(local_peer, NeighborWalkState::GetHandshakesFinish);
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(conn: &DBConn, nk: &NeighborKey) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(|_e| net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+        
+        for slot in slots {
+            let peer_opt = PeerDB::get_peer_at(conn, nk.network_id, slot)
+                .map_err(|_e| net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
```

Comment:
Is this a possible code path? that is, if this function is called, it means that `try_insert_peer` failed -- all of the slots are full with other peers. Could a peer have been removed from db in between? If so, should this function return this slot as the slot to fill?

Reply:
Follow up: why even check for a peer existing here? Couldn't you just return the first slot?